### PR TITLE
Add search endpoint for schedules for scheduler

### PR DESCRIPTION
**Pull request by jpbruinsslot**
### Changes

Added `/schedules/search` endpoint to the scheduler api

### QA notes

Added an extra endpoint to the scheduler that allows more complex queries to be made to retrieve schedules.

1. Go through onboarding
2. Start a scan
3. Going to query the `/schedules/search` to check if we can use more complex queries to get a schedule out
4. 
```console
curl -X POST http://localhost:8004/schedules/search -H "Content-Type: application/json" -d \
'{
  "filters": [
    {
      "column": "data",
      "field": "boefje__id",
      "operator": "eq",
      "value": "dns-zone"
    }
  ]
}'
```
5. Should yield a schedule, if it doesn't try a different boefje id
---

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [x] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

- [x] Tickets have been created for newly discovered issues.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [ ] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: mula/scheduler/server/handlers/schedules.py [hunk 1]
```diff
@@ -246,3 +255,34 @@ def patch(self, schedule_id: uuid.UUID, schedule: serializers.SchedulePatch) ->
...
+        except storage.errors.StorageError as exc:
+            raise fastapi.HTTPException(
+                status_code=fastapi.status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail=f"error occurred while accessing the database [exception: {exc}]",
+            ) from exc
+        except Exception as exc:
+            self.logger.exception(exc)
+            raise fastapi.HTTPException(
+                status_code=fastapi.status.HTTP_500_INTERNAL_SERVER_ERROR, detail="failed to search schedules"
+            ) from exc
```
[the remark this answers is not in the document]
Comment (jpbruinsslot) [reply mula/scheduler/server/handlers/schedules.py:288]: Yeah I'll pick this up with a refactor of the API of the scheduler

[the remark this answers is not in the document]
Comment (jpbruinsslot) [reply mula/scheduler/server/handlers/schedules.py:288]: https://github.com/minvws/nl-kat-coordination/issues/3162

File: mula/scheduler/server/handlers/schedules.py [hunk 2]
```diff
@@ -246,3 +255,36 @@ def patch(self, schedule_id: uuid.UUID, schedule: serializers.SchedulePatch) ->
             ) from exc
 
         return updated_schedule
+
+    def search(
+        self,
+        request: fastapi.Request,
+        offset: int = Query(0),
+        limit: int = Query(10),
```
Comment (ammar92) [inline mula/scheduler/server/handlers/schedules.py:263]: As far as I know fastapi handles arguments as query parameters by default. Body parameters are usually explicitly specified

```suggestion
        offset: int = 0,
        limit: int = 10,
```

File: mula/scheduler/server/handlers/schedules.py [hunk 3]
```diff
@@ -246,3 +255,36 @@ def patch(self, schedule_id: uuid.UUID, schedule: serializers.SchedulePatch) ->
...
+        filters: storage.filters.FilterRequest | None = Body(...),
+    ) -> Any:
```
Comment (ammar92) [inline mula/scheduler/server/handlers/schedules.py:265]:
```suggestion
    ) -> PaginatedResponse:
```

File: mula/scheduler/server/handlers/schedules.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ammar92) [reply mula/scheduler/server/handlers/schedules.py:288]: I suggest using an `exception_handler()` on the app that specifically acts on these more generic errors (such as `StorageError`). The handler in turn can return an appropriate message and status code

File: mula/scheduler/server/handlers/schedules.py (hunk 2 above)
Comment (jpbruinsslot) [reply mula/scheduler/server/handlers/schedules.py:263]: Cool, thought I had to explicitly specify it.

**Comment by Donnype** [comment]
## Checklist for QA:

- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
  - [ ] I have managed to pass the onboarding flow
  - [x] Objects and Findings are created properly
  - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [x] I checked the logs for errors and/or warnings and made issues where necessary

### What works:

(In total there were 7 schedules in my database.)

- I can use the search endpoint with `==`:
Scherm­afbeelding 2024-10-22 om 16 15 53

Gives the right schedules: 
Scherm­afbeelding 2024-10-22 om 16 18 53

Also using `like`:

image

Gives:
Scherm­afbeelding 2024-10-22 om 16 18 28


### What doesn't work:

- I noticed that giving nonsense values is not handled in the API yet 😁:
image

Scherm­afbeelding 2024-10-22 om 16 23 44

Can be a separate issue though since this might affect more generic filter logic!

### Bug or feature?:

- _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._

**Review by underdarknl — APPROVED** [review]

**Pull request merged**